Fix: expanded postings can cache wrong data when queries are issued "in the future"

alanprot opened the pull request:
**What this PR does**:

See: https://github.com/cortexproject/cortex/issues/6556

This PR includes:
* Added unit test to catch the issue
* Improve fuzz test to add GetSeries call and run queries at a "future" timestamp
* Increase the runs from 100 to 300 on the fuzz test (even though 100 was pretty consistent to catch the issue)
* Bypass caching logic if the query is "in the future"

Additionally, we improved the prom to include all the matchers types: https://github.com/cortexproject/promqlsmith/pull/142

**Which issue(s) this PR fixes**:
Fixes https://github.com/cortexproject/cortex/issues/6556

**Checklist**
- [X] Tests updated
- [NA] Documentation added
- [NA] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

yeya24 commented on pkg/storage/tsdb/expanded_postings_cache.go:
```diff
@@ -140,6 +141,11 @@ type blocksPostingsForMatchersCache struct {
 	seedByHash *seedByHash
 }
 
+func (c *blocksPostingsForMatchersCache) Clear() {
```
Is this only used in tests?

alanprot replied:
for now yeah.. u wanna me to add a comment there?

yeya24 reviewed (commented):
> Improve fuzz test to add GetSeries call and run queries at a "future" timestamp

I don't see we added this to fuzz test?

alanprot commented:
> > Improve fuzz test to add GetSeries call and run queries at a "future" timestamp
> 
> I don't see we added this to fuzz test?

My bad.. i added to the unit test.. lemme add some on the fuzz test as well.

alanprot commented:
@yeya24  Added the get series fuzzy testing.

yeya24 reviewed (approved):
Thanks
